### PR TITLE
feat: add property crud operations

### PR DESCRIPTION
**Pull request by Muhindo-Galien**
closes #ISSUE-NO

work on the property's API
## Tasks

- [x] User can post a property
- [x] user can update a property
- [x] user can delete a property
- [x] user can get a signle propery by property_id
- [x] user can get properties

### CODE REVIEW

**Comment by Muhindo-Galien** [comment]
LGTM

**Comment by Muhindo-Galien** [comment]
LGTM

**Comment by Muhindo-Galien** [comment]
LGTM

File: controllers/properties.js [hunk 1]
```diff
@@ -0,0 +1,181 @@
+const Property = require('../models/Property');
+const {
+  CREATED,OK, UNAUTHORIZED
+} = require('../constants/statusCodes');
+
+const logger = require('../util/logger');
+
+// create property
+
+exports.propertyCreate = async(req,res) => {
```
Comment (desirekaleba) [inline controllers/properties.js]: Use `createProperty` as the function name.

File: controllers/properties.js [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    status,
+    price,
+    state,
+    city,
+    address,
+    type,
+    image_url,
+  } = req.body
+
+  const nweProperty = new Property({
```
Comment (desirekaleba) [inline controllers/properties.js]: `newProperty` not `nweProperty`

File: controllers/properties.js [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    owner: req.user.id,
+    status,
+    price,
+    state,
+    city,
+    address,
+    type,
+    image_url, });
+  console.log(req.user)
```
Comment (desirekaleba) [inline controllers/properties.js]: No `console.log`

File: controllers/properties.js [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+      status: 'success',
+      data: saveProperty });
+  }catch(error){
+    logger.error(error.message);
+  }
+
+}
+//updated property
+
+exports.propertyUpdate = async(req, res) => {
```
Comment (desirekaleba) [inline controllers/properties.js]: Use `updateProperty` as the function name.

File: controllers/properties.js [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+  }catch(err){
+    logger.error(err.message);
+  }
+
+
+}
+
+// deleted property
+
+exports.propertyDelete = async(req, res) => {
```
Comment (desirekaleba) [inline controllers/properties.js]: This too should be `deleteProperty`

File: controllers/properties.js [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+      message: 'Property successfully deleted!'
+    });
+  }catch(err){
+    logger.error(err.message);
+  }
+
+}
+
+// get all properties & a signle property by type
+exports.propertyGetAll = async(req, res) => {
```
Comment (desirekaleba) [inline controllers/properties.js]: `getAllPropertiesByType` as function name.

File: controllers/properties.js [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+    });
+  }catch(err){
+    logger.error(err.message);
+  }
+
+}
+
+// get get one property
+
+exports.propertyGetOne = async(req, res) => {
```
Comment (desirekaleba) [inline controllers/properties.js]: `getPropertyById` as function name.

File: controllers/properties.js [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+//       logger.error(err.message);
+//     }
+//   }else{
+//     res.status(UNAUTHORIZED).json({
+//       status:'error',
+//       message: 'You can only delete your own property'
+//     })
+//   }
+
+// }
```
Comment (desirekaleba) [inline controllers/properties.js]: Removed unused code. And add logic for getting all properties without any filter. use `getAllProperties` as the function name.

File: middlewares/authToken.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
+const { FORBIDDEN,UNAUTHORIZED } = require('../constants/statusCodes');
+const jwt = require('jsonwebtoken');
+
+exports.authToken = (req, res, next) => {
```
Comment (desirekaleba) [inline middlewares/authToken.js]: Use `isAuthenticated` as your middleware name.

File: middlewares/authToken.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  const authHeader = req.headers['authorization'];
+  const token = authHeader && authHeader.split(' ')[1];
+  if(token === null){
+    res.status(UNAUTHORIZED).json({
+      status: 'error',
+      message: 'No access allowed.',
+    });
+  }
+
+  jwt.verify(token, process.env.JWT_SECRET_KEY, (err,user) => {
```
Comment (desirekaleba) [inline middlewares/authToken.js]: Use the token until file to verify your token. Precisely the `compare` method.

Comment (desirekaleba) [reply middlewares/authToken.js]: update it to bind the user to your request.

File: models/Property.js [hunk 11]
```diff
@@ -0,0 +1,19 @@
+const mongoose = require('mongoose');
+const PropertySchema = new mongoose.Schema({
+  owner : {
+    type:mongoose.Schema.Types.ObjectId,
+    required: true
+  },
+  status : String,
+  price : String,
```
Comment (desirekaleba) [inline models/Property.js]: price should be a floating number.

File: routes/property.js [hunk 12]
```diff
@@ -0,0 +1,20 @@
+const { propertyCreate,propertyUpdate,propertyGetOne,propertyGetAll, propertyDelete } = require('../controllers/properties');
```
Comment (desirekaleba) [inline routes/property.js]: Update these accordingly.

File: routes/property.js [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+const { authToken } = require('../middlewares/authToken')
+
+
+router.route('/').post(authToken, propertyCreate);
+
+router.route('/:id').delete(authToken,propertyDelete);
+
+router.route('/:propertyid/sold').patch(authToken,propertyUpdate);
+
+router.route('/').get(propertyGetAll);
```
Comment (desirekaleba) [inline routes/property.js]: This is getting by type. create `getAllProperties` and use it here.

Comment (desirekaleba) [reply routes/property.js]: You should add filtering by type too.

File: server.js [hunk 14]
```diff
@@ -23,6 +23,8 @@ app.use(express.urlencoded({ extended: false }));
 app.use(cors());
 app.use('/api/v1', routes);
 
+
+
```
Comment (desirekaleba) [inline server.js:24]: Remove this blank space.

**Review by desirekaleba — CHANGES_REQUESTED** [review]
Check all my comments and update the code accordingly. Once you finish, push your updates and rerequest the reviewer again.

File: routes/flag.js [hunk 15]
```diff
@@ -0,0 +1,10 @@
+const { createFlag,deleteFlag } = require('../controllers/flags');
+const { isAuthenticated } = require('../middlewares/isAuthenticated');
+
+
+const express = require('express');
+const router = express.Router();
+router.route('/flag/:propertyId').post(isAuthenticated,createFlag);
+router.route('/flag/:propertyId').delete(isAuthenticated,deleteFlag);
```
Comment (desirekaleba) [inline routes/flag.js:8]: :flagId

**Review by desirekaleba — APPROVED** [review]
I'll merge but still need some changes. I'll discuss with you those changes as the project grows

**Pull request merged**